File: leetcode/101-200/144-binary-tree-preorder-traversal/bench/preorder.c
#include <stdio.h>
#include <stdlib.h>

int main(void) {
    long n = 50000, passes = 400;

    long *val   = malloc(n * sizeof(long));
    long *left  = malloc(n * sizeof(long));
    long *right = malloc(n * sizeof(long));

    long state = 12345;
    for (long i = 0; i < n; i++) {
        state = (state * 1103515245L + 12345L) & 2147483647L;
        val[i] = state >> 16;
        left[i] = -1;
        right[i] = -1;
    }

    for (long i = 1; i < n; i++) {
        long cur = 0;
        int placed = 0;
        while (!placed) {
            if (val[i] < val[cur]) {
                if (left[cur] < 0) { left[cur] = i; placed = 1; }
                else cur = left[cur];
            } else {
                if (right[cur] < 0) { right[cur] = i; placed = 1; }
                else cur = right[cur];
            }
        }
    }

    long *stack = malloc(n * sizeof(long));
    long sink = 0;
    for (long p = 0; p < passes; p++) {
        long idx = p % n;
        val[idx] += 1;

        long sp = 0;
        stack[sp++] = 0;
        long pos = 0, acc = 0;
        while (sp > 0) {
            long node = stack[--sp];
            acc += val[node] * (pos + 1);
            pos++;
            long r = right[node], l = left[node];
            if (r >= 0) stack[sp++] = r;
            if (l >= 0) stack[sp++] = l;
        }
        sink += acc;
    }
    printf("%ld\n", sink);
    free(val); free(left); free(right); free(stack);
    return 0;
}

File: leetcode/101-200/144-binary-tree-preorder-traversal/bench/preorder.rs
fn main() {
    let n: i64 = 50000;
    let passes: i64 = 400;
    let nu = n as usize;

    let mut val = vec![0i64; nu];
    let mut left = vec![-1i64; nu];
    let mut right = vec![-1i64; nu];

    let mut state: i64 = 12345;
    for i in 0..nu {
        state = (state * 1103515245 + 12345) & 2147483647;
        val[i] = state >> 16;
    }

    for i in 1..nu {
        let mut cur = 0usize;
        loop {
            if val[i] < val[cur] {
                if left[cur] < 0 { left[cur] = i as i64; break; }
                else { cur = left[cur] as usize; }
            } else {
                if right[cur] < 0 { right[cur] = i as i64; break; }
                else { cur = right[cur] as usize; }
            }
        }
    }

    let mut stack = vec![0i64; nu];
    let mut sink: i64 = 0;
    for p in 0..passes {
        let idx = (p % n) as usize;
        val[idx] += 1;

        let mut sp = 0usize;
        stack[sp] = 0;
        sp += 1;
        let mut pos: i64 = 0;
        let mut acc: i64 = 0;
        while sp > 0 {
            sp -= 1;
            let node = stack[sp] as usize;
            acc += val[node] * (pos + 1);
            pos += 1;
            let r = right[node];
            let l = left[node];
            if r >= 0 { stack[sp] = r; sp += 1; }
            if l >= 0 { stack[sp] = l; sp += 1; }
        }
        sink += acc;
    }
    println!("{}", sink);
}
